lib/events.js

```diff
@@ -351,6 +351,54 @@ EventEmitter.prototype.prependOnceListener =
       return this;
     };
 
+// An awaitable one-time event listener. Registers a once handler and
+// returns an associated promise.
+EventEmitter.prototype.when =
+    function when(type, options = { prepend: false }) {
+      if (options == null || typeof options !== 'object') {
+        // This must be lazy evaluated
+        const errors = require('internal/errors');
+        throw new errors.TypeError('ERR_INVALID_ARG_TYPE',
+                                   'options', 'object');
+      }
+      // Do not pull these in at the top, they have to be lazy evaluated
+      const {
+        createPromise,
+        promiseReject,
+        promiseResolve
+      } = process.binding('util');
+      const promise = createPromise();
+      let resolved = false;
+      const listener = (...args) => {
+        resolved = true;
+        this.removeListener(type, listener);
+        promiseResolve(promise, { emitter: this, args });
+      }
```

Comment:
Missing semicolon here and below for `cancel`?

Reply:
doh lol